session/session.go

```diff
@@ -98,7 +98,7 @@ func newZKSession(servers string, recvTimeout time.Duration, logger stdLogger, c
 	var event zookeeper.Event
 	select {
 	case event = <-events:
-		if event.State != zookeeper.STATE_CONNECTED {
+		if event.State == zookeeper.STATE_AUTH_FAILED || event.State == zookeeper.STATE_EXPIRED_SESSION {
```

Comment:
This still isn't right. You need a loop here that will wait until the `STATE_CONNECTED` event is received. We can't proceed until the connection has been established.